src/core/shared_memory_manager.cc

```diff
@@ -183,7 +183,7 @@ Status
 SharedMemoryManager::UnregisterAllSharedMemory()
 {
   // Serialize all operations that write/read current shared memory regions
-  std::lock_guard<std::mutex> lock(register_mu_);
```

Comment:
You need to lock to protect shared_memory_map_. You can make the body of UnregisterSharedMemory loop a function that is called in multiple places to avoid duplicating that code.